### ISSUE TITLE
Wrong schema defintion for classes based on Dictionary<string, string> (and any other Dictionary)

### ISSUE THREAD

**Issue by antonGritsenko** [domaindrivendev/Swashbuckle.AspNetCore#675]
This case from here https://github.com/mattfrear/Swashbuckle.AspNetCore.Examples/issues/39.
Version: swashbuckle.aspnetcore.swaggergen 1.0.0 (NuGet)

Class model:
```
    public class PersonRequest : Dictionary<string, string>
    {
        public Title Title { get; set; }
        public int Age { get; set; }
        [Description("The first name of the person")]
        public string FirstName { get; set; }
        public decimal? Income { get; set; }
    }
```
Controller's method:
```
        [HttpPost]
        [Route("api/values/person")]

        [SwaggerResponse(200, typeof(PersonResponse), "Successfully found the person")]
        [SwaggerResponseExample(200, typeof(PersonResponseExample))]
        [SwaggerResponse(500, null, "There was an unexpected error")]
        [SwaggerResponseExample(500, typeof(InternalServerResponseExample))]
        [SwaggerRequestExample(typeof(PersonRequest), typeof(PersonRequestExample), jsonConverter: typeof(StringEnumConverter))]
        [SwaggerResponseHeader(200, "Location", "string", "Location of the newly created resource")]
        [SwaggerResponseHeader(200, "ETag", "string", "An ETag of the resource")]
        [Authorize("Customer")]
        public PersonResponse GetPerson([FromBody]PersonRequest personRequest)
        {
            var personResponse = new PersonResponse { Id = 1, FirstName = "Dave" };
            return personResponse;
        }
```

Operation filter _SwaggerResponseExample_ from here: https://github.com/mattfrear/Swashbuckle.AspNetCore.Examples/blob/master/src/Swashbuckle.AspNetCore.Examples/Examples/ExamplesOperationFilter.cs

Now the bug: Ref parameter of the Schema instance is always null for PersonRequest class in SwaggerResponseExample:
```
var schema = schemaRegistry.GetOrRegister(attr.RequestType);
// schema.Ref will always be null for PersonRequest  type
```
Also schemaRegistry.Definitions doesn't contains schema for PersonRequest.

Everything works fine as soon as removed  : Dictionary<string, string>

According to https://swagger.io/docs/specification/data-models/dictionaries/ we can defined dictionaries as 

> type: object
>    additionalProperties:
>       type: string

but even this doesn't help:
`c.MapType<PeopleRequest>(() => new Schema { Type = "object", AdditionalProperties = new Schema(){ Type = "string"}});`

You can repro this quickly by clone https://github.com/mattfrear/Swashbuckle.AspNetCore.Examples and change 
`public class PersonRequest`
to
`public class PersonRequest : Dictionary<string, object>`
in \test\WebApi\Models\PersonRequest.cs
and put breakpoint to SetRequestModelExamples method (line 32 of \src\Swashbuckle.AspNetCore.Examples\Examples\ExamplesOperationFilter.cs)

**Comment by domaindrivendev** [comment]
Swashbuckle aside, what's the serialization behavior you're expecting with following model:
`public class PersonRequest : Dictionary<string, object>`

If I were to post the following JSON:
```json
{
   "Title":"blah",
   "Foo":"bar"
}
```
the instance of `PersonRequest` will have two dictionary items, one for "Title" and another for "Foo". But, the `Title` property itself will still have the value null. So, the properties you have defined on `PersonRequest` are actually redundant.

In other words, are you deriving from `Dictionary` to support additional properties on top of the pre-defined ones? If so, have you tested that this actually works, because for me it doesn't.

**Comment by domaindrivendev** [comment]
If I'm understanding your use-case correctly, I think you may want to use the `JsonExtensionDataAttribute`. This will be honored by the `JsonSerializer` and by Swashbuckle:

```csharp
public class PersonRequest : Dictionary<string, string>
{
    public Title Title { get; set; }

    ...

    [JsonExtensionData]
    public Dictionary<string, object> AdditionalProperties { get; set; }
}
```

**Comment by antonGritsenko** [comment]
Hm, I think you make this too strict.  Sending dynamic data is really often case, and this need to be supported somehow. I gonna play around this to may be give more detailed suggestion, but for now let's close it to improve stat )

**Comment by domaindrivendev** [comment]
I would suggest you take a moment to google “JsonExtensionData”. This is the correct way to support dynamic data on a JSON payload and it’s supported by Swashbuckle.

**Comment by antonGritsenko** [comment]
@domaindrivendev yeah, I already playing for it. One more question, if you don't mind. Does Swashbuckle support any kind of dynamic data with Dictionary? In other word, can I use Dictionary<string, object> directly or only way to do that is custom class with parameter marked by JsonExtensionData inside?

**Comment by domaindrivendev** [comment]
You can certainly use Dictionary directly as the FromBody parameter. However, SB will describe it as just that - an open JSON object, with no indication of expected properties. If you have a hybrid situation, where there is some pre-defined properties and a capability for additional dynamic values, then JsonExtensionData is the way to go, not just from the SB perspective, but for implementing model binding in general. Is there a reason why you can’t use that?

**Comment by domaindrivendev** [comment]
Unless I’m missing something, that will not be an issue if you use JsonExtensionData on a custom type. I don’t think I can be much more clear that that’s the way to go!

**Comment by antonGritsenko** [comment]
First of all sorry for very basic questions, Swagger is something new for me and I'm not fully understand the concept.

Let's go back to pure Dictionary<string, object>

> You can certainly use Dictionary directly as the FromBody parameter. However, SB will describe it as 
just that - an open JSON object, with no indication of expected properties.

Here we have a problem with IOperationFilter. My goal is to set example to response and request. Action method have OperationFilterContext argument. For Response it pretty simple, we can use Responses of OperationFilterContext  and set examples to Value.Examples. This part work as expected, I can generate an example.  

But with Request it will not work like this. Everything that we have is the SchemaRegistry property of OperationFilterContext . To set example we must have schema for the type (Dictionary<string, object>) registered in Definitions of SchemaRegistry. But this do nothing:
```
        public void Apply(Operation operation, OperationFilterContext context)
        {
            // here schemaRegistry.Definitions.Count == 2 (we have other types)
            var schema = context.SchemaRegistry.GetOrRegister(typeof(Dictionary<string, object>));
            // here schemaRegistry.Definitions.Count still == 2, so we can't use it to set example
        }
```

**Comment by domaindrivendev** [comment]
Parameters of type `Dictionary` will have their schema defined inline rather than by reference to an instance in the global `Definitions` dictionary. Which makes sense if you think about it - you might be using `Dictionary<string, object>` in many different operations, for completely different reasons, and so you wouldn't want to have the same example for all of them.

So, to provide per-operation body examples, you'll need to do something along the following lines:

```csharp
public class ExamplesOperationFilter : IOperationFilter
{
    public void Apply(Operation operation, OperationFilterContext context)
    {
        var methodAndPath = $"{context.ApiDescription.HttpMethod} {context.ApiDescription.RelativePath}";

        switch (methodAndPath)
        {
            case "POST api/values/person":
                {
                    var postParam = operation.Parameters[0] as BodyParameter;
                    postParam.Schema.Example = new
                    {
                        foo = "bar"
                    };
                }
                break;
        }
    }
}
```

**Comment by antonGritsenko** [comment]
This is just perfect! Thank you for detailed sample!
